Index: mt7615x/mt_wifi/embedded/common/cmm_info.c
===================================================================
--- mt7615x.orig/mt_wifi/embedded/common/cmm_info.c
+++ mt7615x/mt_wifi/embedded/common/cmm_info.c
@@ -10690,6 +10690,7 @@
 	RTMP_CHIP_CAP *pChipCap = hc_get_chip_cap(pAd->hdev_ctrl);
 
 	if (pAd->CommonCfg.dbdc_mode) { /* Dual Band */
+#ifdef WIFI_SPECTRUM_SUPPORT
 		if ((CapNode == pChipCap->SpectrumWF0ADC)  || (CapNode == pChipCap->SpectrumWF1ADC)
 			|| (CapNode == pChipCap->SpectrumWF0FIIQ) || (CapNode == pChipCap->SpectrumWF1FIIQ)
 			|| (CapNode == pChipCap->SpectrumWF0FDIQ) || (CapNode == pChipCap->SpectrumWF1FDIQ))
@@ -10698,6 +10699,7 @@
 				 || (CapNode == pChipCap->SpectrumWF2FIIQ) || (CapNode == pChipCap->SpectrumWF3FIIQ)
 				 || (CapNode == pChipCap->SpectrumWF2FDIQ) || (CapNode == pChipCap->SpectrumWF3FDIQ))
 			rfic =  RFIC_5GHZ;
+#endif
 	} else { /* Single Band */
 		if (HcGetRadioChannel(pAd) <= 14)
 			rfic = RFIC_24GHZ;
@@ -10748,6 +10750,7 @@
 	RTMP_CHIP_CAP *pChipCap = hc_get_chip_cap(pAd->hdev_ctrl);
 
 	if (pAd->CommonCfg.dbdc_mode) { /* Dual Band */
+#ifdef WIFI_SPECTRUM_SUPPORT
 		if ((CapNode == pChipCap->SpectrumWF0ADC)  || (CapNode == pChipCap->SpectrumWF1ADC)
 			|| (CapNode == pChipCap->SpectrumWF0FIIQ) || (CapNode == pChipCap->SpectrumWF1FIIQ)
 			|| (CapNode == pChipCap->SpectrumWF0FDIQ) || (CapNode == pChipCap->SpectrumWF1FDIQ))
@@ -10756,6 +10759,7 @@
 				 || (CapNode == pChipCap->SpectrumWF2FIIQ) || (CapNode == pChipCap->SpectrumWF3FIIQ)
 				 || (CapNode == pChipCap->SpectrumWF2FDIQ) || (CapNode == pChipCap->SpectrumWF3FDIQ))
 			rfic = RFIC_5GHZ;
+#endif
 	} else { /* Single Band */
 		if (HcGetRadioChannel(pAd) <= 14)
 			rfic = RFIC_24GHZ;

Index: mt7615x/mt_wifi/os/linux/rt_proc.c
===================================================================
--- mt7615x.orig/mt_wifi/os/linux/rt_proc.c
+++ mt7615x/mt_wifi/os/linux/rt_proc.c
@@ -29,7 +29,7 @@
 #include <linux/kernel.h>
 #include <linux/types.h>
 #include <linux/proc_fs.h>
-#include <asm/uaccess.h>
+#include <linux/uaccess.h>
 
 #include "rt_config.h"
 
Index: mt7615x/mt_wifi/os/linux/rt_linux.c
===================================================================
--- mt7615x.orig/mt_wifi/os/linux/rt_linux.c
+++ mt7615x/mt_wifi/os/linux/rt_linux.c
@@ -803,7 +803,7 @@
 #else
 
 	if (osfd->f_mode & FMODE_CAN_READ) {
-		return __vfs_read(osfd, pDataPtr, readLen, &osfd->f_pos);
+		return kernel_read(osfd, pDataPtr, readLen, &osfd->f_pos);
 #endif
 	} else {
 		MTWF_LOG(DBG_CAT_INIT, DBG_SUBCAT_ALL, DBG_LVL_ERROR, ("no file read method\n"));
@@ -817,7 +817,7 @@
 #if (KERNEL_VERSION(4, 1, 0) > LINUX_VERSION_CODE)
 	return osfd->f_op->write(osfd, pDataPtr, (size_t) writeLen, &osfd->f_pos);
 #else
-	return __vfs_write(osfd, pDataPtr, (size_t) writeLen, &osfd->f_pos);
+	return kernel_write(osfd, pDataPtr, (size_t) writeLen, &osfd->f_pos);
 #endif
 }
 
@@ -1917,7 +1917,7 @@
 				MTWF_LOG(DBG_CAT_INIT, DBG_SUBCAT_ALL, DBG_LVL_ERROR, ("no file write method\n"));
 			}
 #else
-			__vfs_write(file_w, msg, strlen(msg), &file_w->f_pos);
+			kernel_write(file_w, msg, strlen(msg), &file_w->f_pos);
 #endif
 				MTWF_LOG(DBG_CAT_INIT, DBG_SUBCAT_ALL, DBG_LVL_ERROR, ("%s", msg));
 				macAddr += AddrStep;
@@ -1977,7 +1977,7 @@
 					MTWF_LOG(DBG_CAT_INIT, DBG_SUBCAT_ALL, DBG_LVL_ERROR, ("no file write method\n"));
 				}
 #else
-				__vfs_write(file_w, msg, strlen(msg), &file_w->f_pos);
+				kernel_write(file_w, msg, strlen(msg), &file_w->f_pos);
 #endif
 				MTWF_LOG(DBG_CAT_INIT, DBG_SUBCAT_ALL, DBG_LVL_ERROR, ("%s", msg));
 				eepAddr += AddrStep;
@@ -2025,7 +2025,7 @@
 				MTWF_LOG(DBG_CAT_INIT, DBG_SUBCAT_ALL, DBG_LVL_ERROR, ("no file write method\n"));
 			}
 #else
-			__vfs_write(file_w, pBuf, BufLen, &file_w->f_pos);
+			kernel_write(file_w, pBuf, BufLen, &file_w->f_pos);
 #endif
 		}
 


Index: mt7615x/mt_wifi/os/linux/Makefile.mt_wifi_ap
===================================================================
--- mt7615x.orig/mt_wifi/os/linux/Makefile.mt_wifi_ap
+++ mt7615x/mt_wifi/os/linux/Makefile.mt_wifi_ap
@@ -13,7 +13,7 @@
 				-I$(WIFI_DRV_PATH)/mt_wifi/ate/include \
 				-I$(WIFI_DRV_PATH)/mt_wifi/include/mcu
 
-DRV_NAME = mt_wifi
+DRV_NAME = mt7615e
 SRC_DIR = ../mt_wifi
 SRC_EMBEDDED_DIR = ../mt_wifi/embedded
 ifeq ($(CONFIG_SUPPORT_OPENWRT),y)

Index: mt7615x/mt_wifi/embedded/common/ee_flash.c
===================================================================
--- mt7615x.orig/mt_wifi/embedded/common/ee_flash.c
+++ mt7615x/mt_wifi/embedded/common/ee_flash.c
@@ -30,6 +30,7 @@
 
 #include	"rt_config.h"
 #include "hdev/hdev.h"
+#include "linux/mtd/mtd.h"
 
 /*decision flash api by compiler flag*/
 #ifdef CONFIG_PROPRIETARY_DRIVER
@@ -137,7 +138,15 @@
 	ULONG a,
 	ULONG b)
 {
-	flash_read(hdev_ctrl, p, a, b);
+	//flash_read(hdev_ctrl, p, a, b);
+	size_t retlen;
+	struct mtd_info *mtd_info = get_mtd_device_nm("factory");
+	if (IS_ERR(mtd_info) || mtd_info == NULL) {
+		printk("ERROR: failed to find 'factory' mtd partiton\n");
+		return;
+	}
+	mtd_read(mtd_info, a, b, &retlen, p);
+	put_mtd_device(mtd_info);
 }
 
 void RtmpFlashWrite(
@@ -146,11 +155,18 @@
 	ULONG a,
 	ULONG b)
 {
-	flash_write(hdev_ctrl, p, a, b);
+	//flash_write(hdev_ctrl, p, a, b);
+	size_t retlen;
+	struct mtd_info *mtd_info = get_mtd_device_nm("factory");
+	if (IS_ERR(mtd_info) || mtd_info == NULL) {
+		printk("ERROR: failed to find 'factory' mtd partiton\n");
+		return;
+	}
+	mtd_write(mtd_info, a, b, &retlen, p);
+	put_mtd_device(mtd_info);
 }
 
 
-
 static NDIS_STATUS rtmp_ee_flash_init(PRTMP_ADAPTER pAd, PUCHAR start);
 
 static BOOLEAN rtmp_ee_init_check(struct _RTMP_ADAPTER *ad)

Index: mt7615x/mt_wifi/chips/mt7615.c
===================================================================
--- mt7615x.orig/mt_wifi/chips/mt7615.c
+++ mt7615x/mt_wifi/chips/mt7615.c
@@ -39,13 +39,13 @@
 #endif
 
 #ifdef CONFIG_AP_SUPPORT
-#define DEFAULT_BIN_FILE "/etc_ro/wlan/MT7615E_EEPROM1.bin"
+#define DEFAULT_BIN_FILE "/lib/firmware/mt7615e.1.eeprom.bin"
 #else
 #define DEFAULT_BIN_FILE "/etc/MT7615E_EEPROM1.bin"
 #endif /* CONFIG_AP_SUPPORT */
 
 #ifdef CONFIG_RT_SECOND_CARD
-#define SECOND_BIN_FILE "/etc_ro/wlan/MT7615E_EEPROM2.bin"
+#define SECOND_BIN_FILE "/lib/firmware/mt7615e.2.eeprom.bin"
 #endif /* CONFIG_RT_SECOND_CARD */
 #ifdef CONFIG_RT_THIRD_CARD
 #define THIRD_BIN_FILE "/etc_ro/wlan/MT7615E_EEPROM3.bin"

Index: mt7615x/mt_wifi/os/linux/rt_profile.c
===================================================================
--- mt7615x.orig/mt_wifi/os/linux/rt_profile.c
+++ mt7615x/mt_wifi/os/linux/rt_profile.c
@@ -98,13 +98,13 @@
 #define xdef_to_str(s)   def_to_str(s)
 #define def_to_str(s)    #s
 
-#define FIRST_EEPROM_FILE_PATH	"/etc_ro/Wireless/RT2860/"
-#define FIRST_AP_PROFILE_PATH		"/etc/Wireless/RT2860/RT2860.dat"
+#define FIRST_EEPROM_FILE_PATH	"/lib/firmware/mt7615e.1.eeprom.bin"
+#define FIRST_AP_PROFILE_PATH		"/etc/wireless/mt7615e/mt7615e.1.dat"
 #define FIRST_STA_PROFILE_PATH      "/etc/Wireless/RT2860/RT2860.dat"
 #define FIRST_CHIP_ID	xdef_to_str(CONFIG_RT_FIRST_CARD)
 
-#define SECOND_EEPROM_FILE_PATH	"/etc_ro/Wireless/iNIC/"
-#define SECOND_AP_PROFILE_PATH	"/etc/Wireless/iNIC/iNIC_ap.dat"
+#define SECOND_EEPROM_FILE_PATH	"/lib/firmware/mt7615e.2.eeprom.bin"
+#define SECOND_AP_PROFILE_PATH	"/etc/wireless/mt7615e/mt7615e.2.dat"
 #define SECOND_STA_PROFILE_PATH "/etc/Wireless/iNIC/iNIC_sta.dat"
 
 #define SECOND_CHIP_ID	xdef_to_str(CONFIG_RT_SECOND_CARD)

Index: mt7615x/mt_wifi/include/os/rt_linux.h
===================================================================
--- mt7615x.orig/mt_wifi/include/os/rt_linux.h
+++ mt7615x/mt_wifi/include/os/rt_linux.h
@@ -135,8 +135,8 @@
 #ifdef CONFIG_AP_SUPPORT
 #ifdef RTMP_MAC_PCI
 #if CONFIG_RTPCI_AP_RF_OFFSET == 0x48000
-#define AP_PROFILE_PATH			"/etc/Wireless/iNIC/iNIC_ap.dat"
-#define AP_RTMP_FIRMWARE_FILE_NAME "/etc_ro/Wireless/iNIC/RT2860AP.bin"
+#define AP_PROFILE_PATH			"/etc/wireless/mt7615e/mt7615e.2.dat"
+#define AP_RTMP_FIRMWARE_FILE_NAME "/lib/firmware/mt7615e.2.eeprom.bin"
 #else
 #define AP_PROFILE_PATH			"/etc/Wireless/RT2860AP/RT2860AP.dat"
 #define AP_RTMP_FIRMWARE_FILE_NAME "/etc/Wireless/RT2860AP/RT2860AP.bin"
@@ -153,15 +153,15 @@
 
 #ifdef RTMP_RBUS_SUPPORT
 /* This used for rbus-based chip, maybe we can integrate it together. */
-#define RTMP_FIRMWARE_FILE_NAME		"/etc_ro/Wireless/RT2860AP/RT2860AP.bin"
+#define RTMP_FIRMWARE_FILE_NAME		"/lib/firmware/mt7615e.1.eeprom.bin"
 #define PROFILE_PATH			"/etc/Wireless/RT2860i.dat"
-#define AP_PROFILE_PATH_RBUS		"/etc/Wireless/RT2860/RT2860.dat"
+#define AP_PROFILE_PATH_RBUS		"/etc/wireless/mt7615e/mt7615e.1.dat"
 #define RT2880_AP_DRIVER_VERSION	"1.0.0.0"
 #endif /* RTMP_RBUS_SUPPORT */
 
 #ifdef SINGLE_SKU_V2
-#define SINGLE_SKU_TABLE_FILE_NAME	"/etc_ro/Wireless/RT2860AP/SingleSKU.dat"
-#define BF_SKU_TABLE_FILE_NAME      "/etc_ro/Wireless/RT2860AP/SingleSKU_BF.dat"
+#define SINGLE_SKU_TABLE_FILE_NAME	"/etc/wireless/mt7615e/SingleSKU.dat"
+#define BF_SKU_TABLE_FILE_NAME      "/etc/wireless/mt7615e/SingleSKU_BF.dat"
 #endif /* SINGLE_SKU_V2 */
 
 #endif /* CONFIG_AP_SUPPORT */

Index: mt7615x/mt_wifi/embedded/include/l1profile.h
===================================================================
--- mt7615x.orig/mt_wifi/embedded/include/l1profile.h
+++ mt7615x/mt_wifi/embedded/include/l1profile.h
@@ -1,7 +1,7 @@
 UCHAR l1profile_default[] = {
 "Default\n\
 INDEX0=MT7615\n\
-INDEX0_profile_path=/etc/Wireless/RT2860/RT2860.dat\n\
+INDEX0_profile_path=/etc/wireless/mt7615e/mt7615e.1.dat\n\
 INDEX0_EEPROM_offset=0x0\n\
 INDEX0_EEPROM_size=0x4000\n\
 INDEX0_EEPROM_name=e2p\n\
@@ -11,10 +11,10 @@
 INDEX0_apcli_ifname=apcli\n\
 INDEX0_mesh_ifname=mesh\n\
 INDEX0_nvram_zone=RT2860\n\
-INDEX0_single_sku_path=/etc_ro/Wireless/RT2860AP/SingleSKU.dat\n\
-INDEX0_bf_sku_path=/etc_ro/Wireless/RT2860AP/SingleSKU_BF.dat\n\
+INDEX0_single_sku_path=/etc/wireless/mt7615e/SingleSKU.dat\n\
+INDEX0_bf_sku_path=/etc/wireless/mt7615e/SingleSKU_BF.dat\n\
 INDEX1=MT7615\n\
-INDEX1_profile_path=/etc/Wireless/iNIC/iNIC_ap.dat\n\
+INDEX1_profile_path=/etc/wireless/mt7615e/mt7615e.2.dat\n\
 INDEX1_EEPROM_offset=0x8000\n\
 INDEX1_EEPROM_size=0x4000\n\
 INDEX1_EEPROM_name=e2p\n\
@@ -24,8 +24,8 @@
 INDEX1_apcli_ifname=apclii\n\
 INDEX1_mesh_ifname=meshi\n\
 INDEX1_nvram_zone=RTDEV\n\
-INDEX1_single_sku_path=/etc_ro/Wireless/RT2860AP/SingleSKU.dat\n\
-INDEX1_bf_sku_path=/etc_ro/Wireless/RT2860AP/SingleSKU_BF.dat\n\
+INDEX1_single_sku_path=/etc/wireless/mt7615e/SingleSKU.dat\n\
+INDEX1_bf_sku_path=/etc/wireless/mt7615e/SingleSKU_BF.dat\n\
 INDEX2=MT7615\n\
 INDEX2_profile_path=/etc/Wireless/WIFI3/RT2870AP.dat\n\
 INDEX2_EEPROM_offset=0x4000\n\
