Index: mt7615e/mt_wifi/os/linux/Makefile.mt_wifi_ap
===================================================================
--- mt7615e.orig/mt_wifi/os/linux/Makefile.mt_wifi_ap
+++ mt7615e/mt_wifi/os/linux/Makefile.mt_wifi_ap
@@ -3,12 +3,13 @@
 				-Idrivers/net/wireless/mt_wifi/ate/include
 
 ifeq ($(CONFIG_SUPPORT_OPENWRT),y)
-EXTRA_CFLAGS = -I$(src)/../src/include \
-				-I$(src)/../src/embedded/include \
-				-I$(src)/../src/ate/include
-DRV_NAME = mt7628
-SRC_DIR = ../src/embedded
+DRV_NAME = mt7615e
+SRC_DIR = ../mt_wifi
+SRC_EMBEDDED_DIR = $(SRC_DIR)/embedded
 obj-m += $(DRV_NAME).o
+EXTRA_CFLAGS = -I$(src)/$(SRC_DIR)/include \
+				-I$(src)/$(SRC_EMBEDDED_DIR)/include \
+				-I$(src)/$(SRC_DIR)/ate/include
 else
 EXTRA_CFLAGS = -Idrivers/net/wireless/mt_wifi/include \
 				-Idrivers/net/wireless/mt_wifi/embedded/include \
@@ -750,7 +751,8 @@
 EXTRA_CFLAGS += -DBRCM_256QAM 
 endif
 endif
-EXTRA_CFLAGS += -Werror
+#EXTRA_CFLAGS += -Werror
+EXTRA_CFLAGS += -Wno-error=date-time
 EXTRA_CFLAGS += -Wframe-larger-than=4096
 ifeq ($(CONFIG_HDR_TRANS_TX_SUPPORT),y)
 	EXTRA_CFLAGS += -DHDR_TRANS_TX_SUPPORT -DCUT_THROUGH_FULL_OFFLOAD
@@ -852,7 +854,8 @@
 EXTRA_CFLAGS += -DCONFIG_HW_HAL_OFFLOAD
 EXTRA_CFLAGS += -DCONFIG_FPGA_MODE
 
-EXTRA_CFLAGS += -Werror
+#EXTRA_CFLAGS += -Werror
+EXTRA_CFLAGS += -Wno-error=date-time
 ifeq ($(CONFIG_HDR_TRANS_TX_SUPPORT),y)
         EXTRA_CFLAGS += -DHDR_TRANS_TX_SUPPORT -DCUT_THROUGH_FULL_OFFLOAD
 endif
@@ -939,7 +939,7 @@
 EXTRA_CFLAGS += -DCONFIG_AP_SUPPORT -DSCAN_SUPPORT -DAP_SCAN_SUPPORT
 EXTRA_CFLAGS += -DDOT11_N_SUPPORT -DDOT11N_DRAFT3 -DSTATS_COUNT_SUPPORT -DIAPP_SUPPORT -DDOT1X_SUPPORT
 #EXTRA_CFLAGS += -DRALINK_ATE -DRALINK_QA -DCONFIG_RT2880_ATE_CMD_NEW
-ifeq ($(CONFIG_RA_HW_NAT_WIFI_NEW_ARCH),n)
+ifneq ($(CONFIG_RA_HW_NAT_WIFI_NEW_ARCH),y)
 EXTRA_CFLAGS += -DCONFIG_RA_NAT_NONE
 endif

Index: mt7615e/mt_wifi/os/linux/rt_profile.c
===================================================================
--- mt7615e.orig/mt_wifi/os/linux/rt_profile.c
+++ mt7615e/mt_wifi/os/linux/rt_profile.c
@@ -24,6 +24,7 @@
  */
 
 #include "rt_config.h"
+#include "os/rt_linux.h"
 
 #ifdef RTMP_UDMA_SUPPORT
 #include "rt_udma.h"
@@ -110,14 +110,14 @@
 #define THIRD_CHIP_ID	xdef_to_str(MT_THIRD_CARD)
 
 #else
-#define FIRST_EEPROM_FILE_PATH	"/etc_ro/Wireless/RT2860/"
-#define FIRST_AP_PROFILE_PATH		"/etc/Wireless/RT2860/RT2860.dat"
-#define FIRST_STA_PROFILE_PATH      "/etc/Wireless/RT2860/RT2860.dat"
+#define FIRST_EEPROM_FILE_PATH	"/etc/wireless/mt7615e/"
+#define FIRST_AP_PROFILE_PATH		"/etc/wireless/mt7615e/mt7615e.1.dat"
+#define FIRST_STA_PROFILE_PATH      "/etc/wireless/mt7615e/mt7615e.1.sta.dat"
 #define FIRST_CHIP_ID	xdef_to_str(MT_FIRST_CARD)
 
-#define SECOND_EEPROM_FILE_PATH	"/etc_ro/Wireless/iNIC/"
-#define SECOND_AP_PROFILE_PATH	"/etc/Wireless/iNIC/iNIC_ap.dat"
-#define SECOND_STA_PROFILE_PATH "/etc/Wireless/iNIC/iNIC_sta.dat"
+#define SECOND_EEPROM_FILE_PATH	"/etc/wireless/mt7615e/"
+#define SECOND_AP_PROFILE_PATH	"/etc/wireless/mt7615e/mt7615e.2.dat"
+#define SECOND_STA_PROFILE_PATH "/etc/wireless/mt7615e/mt7615e.2.sta.dat"
 
 #define SECOND_CHIP_ID	xdef_to_str(MT_SECOND_CARD)

Index: mt7615e/mt_wifi/os/linux/rt_proc.c
===================================================================
--- mt7615e.orig/mt_wifi/os/linux/rt_proc.c
+++ mt7615e/mt_wifi/os/linux/rt_proc.c
@@ -29,6 +29,7 @@
 #include <linux/kernel.h>
 #include <linux/types.h>
 #include <linux/proc_fs.h>
+#include <linux/uaccess.h>
 #include <asm/uaccess.h>
 
 #include "rt_config.h"
@@ -62,7 +62,8 @@
 #endif /* PROCREG_DIR */
 
 #ifdef CONFIG_PROC_FS
-extern struct proc_dir_entry *procRegDir;
+//extern struct proc_dir_entry *procRegDir;
+struct proc_dir_entry *procRegDir;
 
 #ifdef VIDEO_TURBINE_SUPPORT
 extern BOOLEAN UpdateFromGlobal;

Index: mt7615e/mt_wifi/os/linux/rt_linux.c
===================================================================
--- mt7615e.orig/mt_wifi/os/linux/rt_linux.c
+++ mt7615e/mt_wifi/os/linux/rt_linux.c
@@ -36,6 +36,7 @@
 #include "rt_os_util.h"
 #include "security/dot11i_wpa.h"
 #include <linux/rtnetlink.h>
+#include "linux/mtd/mtd.h"
 
 #include "rt_os_net.h"
 #ifdef MEM_ALLOC_INFO_SUPPORT
@@ -348,6 +348,19 @@
 
 #endif /* CONFIG_RALINK_FLASH_API */
 
+void flash_read_openwrt(UCHAR *p, ULONG a, ULONG b)
+{
+	size_t retlen;
+	struct mtd_info *mtd_info = get_mtd_device_nm("Factory");
+	if (IS_ERR(mtd_info) || mtd_info == NULL) {
+		printk("ERROR: failed to find 'Factory' mtd partiton\n");
+		return;
+	}
+	mtd_read(mtd_info, a, b, &retlen, p);
+	put_mtd_device(mtd_info);
+
+}
+
 void RtmpFlashRead(
 	UCHAR *p,
 	ULONG a,
@@ -362,12 +375,25 @@
 	//wilsonl, fix compile for temp
 #ifndef MT7622
 	/* ra_mtd_read_nm("Factory", a&0xFFFF, (size_t) b, p); */
-	ra_mtd_read_nm("Factory", a, (size_t) b, p);  /* triple card needs to access beyond 0xFFFF , don't mask */
+	//ra_mtd_read_nm("Factory", a, (size_t) b, p);  /* triple card needs to access beyond 0xFFFF , don't mask */
+	flash_read_openwrt(p, a, b);
 #endif
 #endif
 #endif /* CONFIG_RALINK_FLASH_API */
 }
 
+void flash_write_openwrt(UCHAR *p, ULONG a, ULONG b)
+{
+	size_t retlen;
+	struct mtd_info *mtd_info = get_mtd_device_nm("Factory");
+	if (IS_ERR(mtd_info) || mtd_info == NULL) {
+		printk("ERROR: failed to find 'Factory' mtd partiton\n");
+		return;
+	}
+	mtd_write(mtd_info, a, b, &retlen, p);
+	put_mtd_device(mtd_info);
+}
+
 void RtmpFlashWrite(
 	UCHAR * p,
 	ULONG a,
@@ -382,7 +408,8 @@
 	//wilsonl, fix compile for temp
 #ifndef MT7622
 	/* ra_mtd_write_nm("Factory", a&0xFFFF, (size_t) b, p); */
-	ra_mtd_write_nm("Factory", a, (size_t) b, p);  /* triple card needs to access beyond 0xFFFF , don't mask */
+	//ra_mtd_write_nm("Factory", a, (size_t) b, p);  /* triple card needs to access beyond 0xFFFF , don't mask */
+	flash_write_openwrt(p, a, b);
 #endif
 #endif
 #endif /* CONFIG_RALINK_FLASH_API */
@@ -912,7 +939,7 @@
 		return osfd->f_op->read(osfd, pDataPtr, readLen, &osfd->f_pos);
 #else
 	if (osfd->f_mode & FMODE_CAN_READ) {
-		return __vfs_read(osfd, pDataPtr, readLen,&osfd->f_pos);
+		return kernel_read(osfd, pDataPtr, readLen,&osfd->f_pos);
 #endif
 	} else {
 		MTWF_LOG(DBG_CAT_INIT, DBG_SUBCAT_ALL, DBG_LVL_ERROR, ("no file read method\n"));
@@ -926,7 +953,7 @@
 #if LINUX_VERSION_CODE < KERNEL_VERSION(4, 1, 0)
 	return osfd->f_op->write(osfd, pDataPtr, (size_t) writeLen, &osfd->f_pos);
 #else
-	return __vfs_write(osfd, pDataPtr, (size_t) writeLen, &osfd->f_pos);
+	return kernel_write(osfd, pDataPtr, (size_t) writeLen, &osfd->f_pos);
 #endif
 }
 
Index: mt7615e/mt_wifi/os/linux/rt_rbus_pci_util.c
===================================================================
--- mt7615e.orig/mt_wifi/os/linux/rt_rbus_pci_util.c
+++ mt7615e/mt_wifi/os/linux/rt_rbus_pci_util.c
@@ -21,7 +21,7 @@
 #include "rt_os_util.h"
 
 #if LINUX_VERSION_CODE > KERNEL_VERSION(3,10,20)
-#include <asm-generic/pci-dma-compat.h>
+#include <linux/pci-dma-compat.h>
 #endif
 
 #ifdef RTMP_MAC_PCI

Index: mt7615e/mt_wifi/include/chip/mt7615.h
===================================================================
--- mt7615e.orig/mt_wifi/include/chip/mt7615.h
+++ mt7615e/mt_wifi/include/chip/mt7615.h
@@ -16,7 +16,7 @@
 #define MT7615_PDA_PORT		0xf800
 
 #define MT7615_CT_PARSE_LEN	72
-#define MT7615_BIN_FILE_PATH "/etc/Wireless/WIFI_RAM_CODE_MT7615.bin"
+#define MT7615_BIN_FILE_PATH "/lib/firmware/WIFI_RAM_CODE_MT7615.bin"
 
 #ifdef RF_LOCKDOWN
 #define COUNTRY_CODE_BYTE0_EEPROME_OFFSET   0x11C

Index: mt7615e/mt_wifi/embedded/common/cmm_profile.c
===================================================================
--- mt7615e.orig/mt_wifi/embedded/common/cmm_profile.c
+++ mt7615e/mt_wifi/embedded/common/cmm_profile.c
@@ -5464,8 +5464,8 @@
 
 	/* get MAC address*/
 #ifdef RTMP_FLASH_SUPPORT
-	RtmpFlashRead(&mac_maybe[0][0], 0x40000 + 0x04, MAC_ADDR_LEN);
-	RtmpFlashRead(&mac_maybe[1][0], 0x48000 + 0x04, MAC_ADDR_LEN);
+	RtmpFlashRead(&mac_maybe[0][0], MT_FIRST_IF_RF_OFFSET + 0x04, MAC_ADDR_LEN);
+	RtmpFlashRead(&mac_maybe[1][0], MT_SECOND_IF_RF_OFFSET + 0x04, MAC_ADDR_LEN);
 	MTWF_LOG(DBG_CAT_CFG, DBG_SUBCAT_ALL, DBG_LVL_TRACE, ("mac addr1 in flash=%02x:%02x:%02x:%02x:%02x:%02x!\n", PRINT_MAC(mac_maybe[0])));
 	MTWF_LOG(DBG_CAT_CFG, DBG_SUBCAT_ALL, DBG_LVL_TRACE, ("mac addr2 in flash=%02x:%02x:%02x:%02x:%02x:%02x!\n", PRINT_MAC(mac_maybe[1])));
 #endif /* RTMP_FLASH_SUPPORT */

Index: mt7615e/mt_wifi/embedded/include/eeprom.h
===================================================================
--- mt7615e.orig/mt_wifi/embedded/include/eeprom.h
+++ mt7615e/mt_wifi/embedded/include/eeprom.h
@@ -91,8 +91,8 @@
 #endif /* BB_SOC */
 
 #define EEPROM_DFT_FILE_DIR		"/etc_ro/wlan/"
-#define EEPROM_1ST_FILE_DIR		"/etc_ro/Wireless/RT2860/"
-#define EEPROM_2ND_FILE_DIR		"/etc_ro/Wireless/iNIC/"
+#define EEPROM_1ST_FILE_DIR		"/lib/firmware/"
+#define EEPROM_2ND_FILE_DIR		"/lib/firmware/"
 
 #ifndef EEPROM_DEFAULT_FILE_PATH
 /* RFIC 2820 */
@@ -100,7 +100,7 @@
 
 #ifdef MT7615
 #undef EEPROM_DEFAULT_FILE_PATH
-#define EEPROM_DEFAULT_FILE_PATH                     "/etc_ro/wlan/MT7615_EEPROM.bin"
+#define EEPROM_DEFAULT_FILE_PATH                     "/lib/firmware/mt7615e.1.eeprom.bin"
 #endif /* MT7615 */
 
 
@@ -113,10 +113,12 @@
 #if defined (CONFIG_RT2880_FLASH_32M)
 #define DEFAULT_RF_OFFSET					0x1FE0000
 #else
-#define DEFAULT_RF_OFFSET					0x40000
+//#define DEFAULT_RF_OFFSET					0x40000
+#define DEFAULT_RF_OFFSET					MT_FIRST_IF_RF_OFFSET
 #endif
 
-#define SECOND_RF_OFFSET					0x48000
+//#define SECOND_RF_OFFSET					0x48000
+#define SECOND_RF_OFFSET					MT_SECOND_IF_RF_OFFSET
 #endif /* RTMP_FLASH_SUPPORT */
 
 
Index: mt7615e/mt_wifi/embedded/common_rtmp_init.c
===================================================================
--- mt7615e.orig/mt_wifi/embedded/common/rtmp_init.c
+++ mt7615e/mt_wifi/embedded/common/rtmp_init.c
@@ -2595,8 +2595,8 @@
 /*	if ((IS_PCIE_INF(pAd))) */
 	{
 		/* specific for RT6855/RT6856 */
-		pAd->E2P_OFFSET_IN_FLASH[0] = 0x40000;
-		pAd->E2P_OFFSET_IN_FLASH[1] = 0x48000;
+		pAd->E2P_OFFSET_IN_FLASH[0] = MT_FIRST_IF_RF_OFFSET;
+		pAd->E2P_OFFSET_IN_FLASH[1] = MT_SECOND_IF_RF_OFFSET;
 	}
 #endif /* RTMP_FLASH_SUPPORT */
 #endif /* MULTIPLE_CARD_SUPPORT */

Index: mt7615e/mt_wifi/chips/mt7615.c
===================================================================
--- mt7615e.orig/mt_wifi/chips/mt7615.c
+++ mt7615e/mt_wifi/chips/mt7615.c
@@ -39,13 +39,13 @@
 #endif
 
 #ifdef CONFIG_AP_SUPPORT
-#define DEFAULT_BIN_FILE "/etc_ro/wlan/MT7615E_EEPROM1.bin"
+#define DEFAULT_BIN_FILE "/lib/firmware/mt7615e.1.eeprom.bin"
 #else
 #define DEFAULT_BIN_FILE "/etc/MT7615E_EEPROM1.bin"
 #endif /* CONFIG_AP_SUPPORT */
 
 #ifdef MT_SECOND_CARD
-#define SECOND_BIN_FILE "/etc_ro/wlan/MT7615E_EEPROM2.bin"
+#define SECOND_BIN_FILE "/lib/firmware/mt7615e.2.eeprom.bin"
 #endif /* MT_SECOND_CARD */
 #ifdef MT_THIRD_CARD
 #define THIRD_BIN_FILE "/etc_ro/wlan/MT7615E_EEPROM3.bin"

Index: mt7615e/mt_wifi/embedded/tools/bin2h.c
===================================================================
--- mt7615e.orig/mt_wifi/embedded/tools/bin2h.c
+++ mt7615e/mt_wifi/embedded/tools/bin2h.c
@@ -44,6 +44,7 @@
 		return -1;
     }
 
+    printf("bin2h: input file successfully opened: '%s'\n", infname);
     outfile = fopen(outfname, mode);
 
     if (outfile == (FILE *) NULL) {
@@ -52,6 +53,7 @@
        	return -1;
     }
 
+    printf("bin2h: output file successfully opened: '%s'\n", outfname);
     fputs("/* AUTO GEN PLEASE DO NOT MODIFY IT */ \n",outfile);
     fputs("/* AUTO GEN PLEASE DO NOT MODIFY IT */ \n",outfile);
     fputs("\n",outfile);

Index: mt7615e/mt_wifi/os/linux/rt_main_dev.c
===================================================================
--- mt7615e.orig/mt_wifi/os/linux/rt_main_dev.c
--- mt7615e/mt_wifi/os/linux/rt_main_dev.c
@@ -47,7 +47,7 @@
 
 // TODO: shiang-6590, remove it when MP
 #ifdef RTMP_MAC_PCI
-MODULE_LICENSE("Proprietary");
+MODULE_LICENSE("GPL");
 #endif /* RTMP_MAC_PCI */
 // TODO: End---
 

